components/scheduler/executor_executors.go

```diff
@@ -0,0 +1,493 @@
+	manualStartExecutionDeadline = startWorkflowMaxDeadline
+
+	// Upper bound on how many times starting an individual buffered action should be retried.
+	ExecutorMaxStartAttempts = 10
+
+	errTypeRetryLimitExceeded = "RetryLimitExceeded"
+	errTypeRateLimited        = "RateLimited"
+	errTypeAlreadyStarted     = "serviceerror.WorkflowExecutionAlreadyStarted"
+)
+
+func RegisterExecutorExecutors(registry *hsm.Registry, options ExecutorTaskExecutorOptions) error {
+	e := executorTaskExecutor{
+		ExecutorTaskExecutorOptions: options,
+	}
+	return hsm.RegisterTimerExecutor(registry, e.executeExecuteTask)
+}
+
+func (e executorTaskExecutor) executeExecuteTask(env hsm.Environment, node *hsm.Node, task ExecuteTask) error {
+	schedulerNode := node.Parent
+	scheduler, err := loadScheduler(schedulerNode)
+	if err != nil {
+		return err
+	}
+	logger := newTaggedLogger(e.BaseLogger, scheduler)
+
+	executor, err := e.loadExecutor(node)
+	if err != nil {
+		return err
+	}
+
+	// Make sure we have something to start. If not, we can clear the buffer and
+	// transition to the waiting state.
+	executionInfo := scheduler.Schedule.Action.GetStartWorkflow()
+	if executionInfo == nil || len(executor.BufferedStarts) == 0 {
+		return hsm.MachineTransition(node, func(e Executor) (hsm.TransitionOutput, error) {
+			e.ExecutorInternal.BufferedStarts = nil
+			return TransitionWait.Apply(e, EventWait{
+				Node: node,
+			})
+		})
+	}
+
+	// Drain the Executor's BufferedStarts. drainBuffer will update Executor's
+	// BufferedStarts, as well as Scheduler metadata.
+	e.drainBuffer(logger, env, executor, scheduler, executionInfo)
+
+	// Update SchedulerInfo metadata.
+	err = hsm.MachineTransition(schedulerNode, func(s Scheduler) (hsm.TransitionOutput, error) {
+		s.Info = scheduler.Info
+		return hsm.TransitionOutput{}, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// Find the earliest possible time a BufferedStart can be retried, if any, based
+	// on BackoffTime.
+	var retryAt time.Time
+	for _, start := range executor.BufferedStarts {
+		backoffTime := start.BackoffTime.AsTime()
+
+		// Zero-value protobuf timestamps deserialize to the UNIX epoch, skip if the
+		// backoffTime looks invalid/unset.
+		if backoffTime.Before(start.ActualTime.AsTime()) {
+			continue
+		}
+
+		if retryAt.IsZero() || backoffTime.Before(retryAt) {
+			retryAt = backoffTime
+		}
+	}
+
+	if !retryAt.IsZero() {
+		// When retryAt is set, reschedule.
+		return hsm.MachineTransition(node, func(e Executor) (hsm.TransitionOutput, error) {
+			e.ExecutorInternal.BufferedStarts = executor.BufferedStarts
+			return TransitionExecute.Apply(e, EventExecute{
+				Node:     node,
+				Deadline: retryAt,
+			})
+		})
+	}
+
+	// No more buffered starts, or remaining starts are waiting for a workflow to be
+	// closed. We can transition to waiting.
+	return hsm.MachineTransition(node, func(e Executor) (hsm.TransitionOutput, error) {
+		e.ExecutorInternal.BufferedStarts = executor.BufferedStarts
```

Comment:
Same note as above, you may lose buffered starts if more starts are buffered while you release the lock.